portal/src/graphql/adminapi/AddUserScreen.tsx

```diff
@@ -0,0 +1,271 @@
+import { Label, PrimaryButton, Text, TextField } from "@fluentui/react";
+
+import { useAppConfigQuery } from "../portal/query/appConfigQuery";
+import NavBreadcrumb, { BreadcrumbItem } from "../../NavBreadcrumb";
+import ShowLoading from "../../ShowLoading";
+import ShowError from "../../ShowError";
+import NavigationBlockerDialog from "../../NavigationBlockerDialog";
+import PasswordField, {
+  extractGuessableLevel,
+  isPasswordValid,
+} from "../../PasswordField";
+import { useTextField } from "../../hook/useInput";
+import { PasswordPolicyConfig, PortalAPIAppConfig } from "../../types";
+
+import styles from "./AddUserScreen.module.scss";
+
+interface AddUserContentProps {
+  appConfig: PortalAPIAppConfig | null;
+}
+
+interface AddUserScreenState {
+  name: string;
+  username: string;
+  email: string;
+  phone: string;
+  password: string;
+}
+
+interface FieldVisible {
+  username: boolean;
+  email: boolean;
+  phone: boolean;
+  password: boolean;
+}
+
+interface AddUserViolation {
+  type: "OneOfLoginIdType" | "PasswordInvalid";
+}
+
+function checkOnlyOneOfLoginIdFieldFilled(screenState: AddUserScreenState) {
+  return (
+    [screenState.username, screenState.email, screenState.phone].filter(
+      (fieldValue) => fieldValue.length > 0
+    ).length === 1
+  );
+}
+
+function validate(
+  screenState: AddUserScreenState,
+  isFieldVisible: FieldVisible,
+  passwordPolicy: PasswordPolicyConfig
+) {
+  const errors: AddUserViolation[] = [];
+  if (!checkOnlyOneOfLoginIdFieldFilled(screenState)) {
+    errors.push({ type: "OneOfLoginIdType" });
+  }
+  if (isFieldVisible.password) {
+    const passwordCheckResult = zxcvbn(screenState.password);
+    const guessableLevel = extractGuessableLevel(passwordCheckResult);
+    if (
+      !isPasswordValid(passwordPolicy, screenState.password, guessableLevel)
+    ) {
+      errors.push({ type: "PasswordInvalid" });
+    }
+  }
+  return errors;
+}
+
+function selectViolations(
+  violations: AddUserViolation[],
+  violationType: AddUserViolation["type"]
+) {
+  return violations.filter((violation) => violation.type === violationType);
+}
+
+const AddUserContent: React.FC<AddUserContentProps> = function AddUserContent(
+  props: AddUserContentProps
+) {
+  const { appConfig } = props;
+  const { renderToString } = useContext(Context);
+
+  const [violations, setViolations] = useState<AddUserViolation[]>([]);
+
+  const isFieldVisible = useMemo(() => {
+    const loginIdKeys = appConfig?.identity?.login_id?.keys ?? [];
+    // We consider them as enabled if they are listed as allowed login ID keys.
+    const username = loginIdKeys.find((key) => key.type === "username") != null;
+    const email = loginIdKeys.find((key) => key.type === "email") != null;
+    const phone = loginIdKeys.find((key) => key.type === "phone") != null;
+
+    const primaryAuthenticators =
+      appConfig?.authentication?.primary_authenticators ?? [];
+    const password =
+      primaryAuthenticators.length > 0
```

Comment:
@axolotl0212 sorry the condition is actually quite complex:
- if only `password` in `primary_authenticators`, all three login ID type need password.
- if only `oob_otp` in `primary_authenticators`, only email & phone is possible, and no password is needed.
- if `primary_authenticators` has both `oob_otp` & `password`, and `oob_otp` is before `password`, email & phone does not need password, but username need password.
- if `primary_authenticators` has both `oob_otp` & `password`, and `oob_otp` is after `password`, all three login ID type need password.

This need design update, so we can keep it as-is for now.

cc @louischan-oursky @keithtkl

Reply:
@kiootic @louischan-oursky What do you think if we simply take out password field for the case that if only `oob_otp` in `primary_authenticators` while other cases still requires user to input the password when add an user?

Reply:
I guess we would disable the password field in cases where password is not needed.